cachi2/core/package_managers/gomod.py

```diff
@@ -660,228 +669,273 @@ def _should_vendor_deps(
     return False, False
 
 
-def _get_golang_version(
-    module_name: str,
-    app_dir: RootedPath,
-    commit_sha: Optional[str] = None,
-    update_tags: bool = False,
-) -> str:
-    """
-    Get the version of the Go module in the input Git repository in the same format as `go list`.
-
-    If commit doesn't point to a commit with a semantically versioned tag, a pseudo-version
-    will be returned.
-
-    :param module_name: the Go module's name
-    :param app_dir: the path to the module directory
-    :param commit_sha: the Git commit SHA1 of the Go module to get the version for
-    :param update_tags: determines if `git fetch --tags --force` should be run before
-        determining the version. If this fails, it will be logged as a warning.
-    :return: a version as `go list` would provide
-    :raises FetchError: if failed to fetch the tags on the Git repository
-    """
-    # If the module is version v2 or higher, the major version of the module is included as /vN at
-    # the end of the module path. If the module is version v0 or v1, the major version is omitted
-    # from the module path.
-    module_major_version = None
-    match = re.match(r"(?:.+/v)(?P<major_version>\d+)$", module_name)
-    if match:
-        module_major_version = int(match.groupdict()["major_version"])
-
-    repo = git.Repo(app_dir.root)
-    if update_tags:
+class ModuleVersionResolver:
+    """Resolves the versions of Go modules in a git repository."""
+
+    def __init__(self, repo_path: RootedPath):
+        """Initialize a ModuleVersionResolver for the repo located at repo_path."""
+        self._repo = git.Repo(repo_path)
+        self._commit = self._repo.commit(self._repo.rev_parse("HEAD").hexsha)
+        self._fetch_remote_tags()
```

Comment:
I'm generally not a fan of `__init__`s that do too much, especially network IO.

And in this case, `_fetch_remote_tags` also has implications for build reproducibility, since the same repo+ref can get a different main module version on the next rebuild. Which is fine, it's always been that way, but I'd like to keep it explicit (it was at least a param before).

Maybe something like this
```python
version_resolver = ModuleVersionResolver(repo_path)
version_resolver.fetch_new_tags()

# or
version_resolver = ModuleVersionResolver.for_repo(repo_path, fetch_new_tags=True)
```

Reply:
+1 to removing the network call from `__init__`.

On the other hand, from a pure design perspective, one could instantiate the object and call `get_golang_version` without ever calling `fetch_remote_tags`, which would be a problem if done by accident.

Still, I can't think of a better way of handling this, maybe logging a warning if that happens? I don't think it's a major issue though, so probably just leave it as it is.